text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,184 @@ are always executed in WSL.
 separate TOML config section). This is due to the fact that the ``watch``
 command conceptually monitors the schema file changes and automatically
-creates and applied migrations. So after the schema file changes are applied
-to the database, the ``migration.apply.after`` hook will be executed.
+creates and applied migrations. So the schema file changes would trigger
+``migration.apply.before``, ``schema.update.before``,
+``migration.apply.after``, and ``schema.update.after`` hooks in that relative
+order.
+
+
+Watch Configuration
+===================
+
+Sometimes in a Gel project there's a need to respond to some file changes: run
+a migration due to schema change or run code generators due to query file
+change. The CLI already supports watching the schema for changes and applying
+them, but we can generalize this mechanism and make it more flexible, much
+like the hooks that respond to CLI commands.
+
+The idea is to have a mechanism for specifying what changes you want to watch
+and a script that gets triggered by those changes. Unlike hooks, there are no
+"before" and "after" triggers here. Only one kind of response is possible:
+trigger when a change to the watched entity is detected.
+
+The watch configuration will be described in the ``gel.toml`` project manifest
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts. Potentially, we will
+also later introduce a ``[watch.gel-config]`` sub-table for monitoring changes
+to the various database config values and responding to them.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+The ``gel watch`` command (without needing further options) would then be used
+to start the watch process and monitor whatever is specified in the
+``gel.toml``. By default ``gel watch`` will only watch the files specified in
+the ``gel.toml`` config and execute the trigger scripts. Running ``gel watch
+--migrate`` will additionally monitor schema changes and perform real-time
+migrations.
+
+This is a backwards incompatible change compared to how ``gel watch`` operates
+now.
+
+The output of the scripts will then appear in the same TTY as the ``gel
+watch`` command. We may also consider setting up a logfile (because that
+creates a record which survives closing of terminals or reboots) as an
+additional convenience feature. This can be specified in the general
+``[watch]`` section as ``logfile="<path-to-logfile>"``.
+
+We may want to setup a debouncer so that we can delay before triggering the
+scripts on a sequence of changes. This is mostly to reduce unnecessary
+multiple triggers for the same watched entity. The debouncer delay may be
+configured in the ``[watch]`` section as ``deboucer-delay=<integer>`` with the
+value being the delay in milliseconds.
```

Comment:
Debouncing seems like a must-have feature, but being able to tweak the timing seems like a nice-to-have one.